projects/hathor-wallet-headless/0004-tx-transaction.md

```diff
@@ -0,0 +1,209 @@
+
+`shuffle` action arguments:
+
+- `target`: 'outputs', 'inputs' or 'all', defaults to 'all'.
+
+`add-output` action arguments:
+
+- `output`: an output instruction (cannot be an `action`)
+
+`add-input` action arguments:
+
+- `input`: an input instruction (cannot be an `action`)
+
+`fill-change` does not have arguments, as the name implies it will check the transaction balance and find any change outputs necessary.
+
+### Example
+
+The following template will mint 0.01 TST, generate another mint authority to the wallet and it will have a 'foobar' data output as the first output.
+The mint deposit inputs are manually added to the transaction and all outputs are shuffled except for the data output at position 0.
+
+```json
+{
+  "inputs": [
+    { "fill": 2, "token": "00" },
+    { "fill": 1, "token": "TST", "authority": "mint" }
+  ],
+  "outputs": [
+    { "type": "token", "token": "TST", "amount": 1 },
+    { "type": "token", "token": "TST", "authority": "mint", "amount": 1 },
+    { "type": "action", "action": "shuffle" },
+    { "type": "data", "data": "foobar", "position": 0 },
+  ],
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The sequence of operations for the template is very straightforward, we choose the inputs, then we form the output array and once the transaction is formed we can perform actions on it to finalize the transaction instance.
+
+## Building the template
+
+### Selecting inputs
+
+If the instruction type is `utxo` we need to use the wallet's storage `selectUtxos` method to find utxos, then we can add them to the inputs array, if any change is required and `autochange` is true or missing we add them to the outputs.
+
+#### Token selection
+
+The following instruction will find 0.10 TST tokens from address `Wtest` and add them to the inputs, adding also change outputs if necessary.
+
+```json
+{
+  "token": "TST",
+  "fill": 10,
+  "address": "Wtest",
+}
+```
+
+This translates to the following query:
+
+```ts
+wallet.storage.selectUtxos({
+  token: "TST",
+  target_amount: 10,
+  filter_address: "Wtest",
+  only_available_utxos: true,
+});
+```
+
+All utxos found will have at least 0.10 TST, they will be unlocked and from address `Wtest`.
+
+#### Authority selection
+
+The following instruction will find 1 TST Mint authority utxo and add them to the inputs.
+
+```json
+{
+  "token": "TST",
+  "fill": 1,
+  "authority": "mint",
+}
+```
+
+This translates to the following query:
+
+```ts
+wallet.storage.selectUtxos({
+  token: "TST",
+  target_amount: 1,
+  authorities: 1, // MINT_AUTHORITY
+  only_available_utxos: true,
+});
+```
+
+We will find 1 unlocked mint authority for TST.
+
+## Template implementation
+
+The `TransactionTemplate` class can be initiated from an object (the JSON template specified in this design) or it can be initialized empty and we add each input/output/action instruction with methods of the class.

```

Comment:
thought: This is a really nice feature to have, especially for those who already know for a good use case to use it. However thinking from the perspective of a curious reader, the applications are not obvious.

Since this text will be at the public RFCs, would it be possible to add a use case example to better illustrate its usage?

Reply:
We have an example in the guide-level for a mint with data output.
I could add more examples, for instance simple mint/melt, send transaction with data output, multiple custom tokens, etc.
What do you think?

Reply:
I think this would be great! It would give a better idea of the applications to the reader.

Reply:
Closing as the design changed since this comment